### PR TITLE
Commenting rate limit for users with low karma

### PR DESCRIPTION
**Pull request by s-cheng**
This PR sets up a commenting rate limit (currently this is 4 comments per 30 min) for users with karma under a configurable amount (EA Forum will set it at 30). We also track when a user has submitted a comment that makes them hit a rate limit, to help us track how disruptive these are. [Figma here](https://www.figma.com/file/Y9VGc5yC2RqKCYS6N4cSaN/Q1.2-2023?node-id=991-31178&t=4XCbfGZQZ4A1DxAM-0)

There are also a few UI updates (some of which affect LW):
1. Added a yellowish warning banner for the rate limit message
2. Changed the "restore autosaved text" banner from red to primary colored, plus made it dismissible
3. When the user is commenting rate limited, the "new comment" form does appear now, but the "Submit" button is disabled until the rate limit has expired
4. Added a "Comments #" header for the comments section, and removed the "New comment" line from the form, for EA Forum only
5. Changed the comment form placeholder text, for EA Forum only
6. Changed the cancel & submit button designs and copy, for EA Forum only

Light mode EAF:
Screen Shot 2023-04-25 at 11 32 36 PM

-----
Dark mode EAF:
Screen Shot 2023-04-25 at 11 33 04 PM

-----
Light mode LW:
Screen Shot 2023-04-25 at 5 42 05 PM

-----
Dark mode LW:
Screen Shot 2023-04-25 at 5 41 44 PM

-----
For the EA Forum, the new comment box looks a lot cleaner now:
Screen Shot 2023-04-25 at 11 54 52 PM

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204474949146164) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/questions/RelatedQuestionsList.tsx [hunk 1]
```diff
@@ -41,6 +41,8 @@ const RelatedQuestionsList = ({ post, classes }: {
   
   const showParentLabel = sourcePostRelations.length > 0
   const showSubQuestionLabel = (sourcePostRelations.length > 0) && (targetPostRelations.length > 0)
+  
+  if (!totalRelatedQuestionCount) return null
```
Comment (s-cheng) [inline packages/lesswrong/components/questions/RelatedQuestionsList.tsx:45]: This component was adding a bunch of extra margin, and I think "related questions" are used ~never, so I figured this was fine.

**Comment by agnesstenlund** [comment]
I really like this 😍 thanks for the speedy turnaround! A few things:

- [ ] I never see the Cancel button 🤔 when should it show?
- [ ]  I also don't get the limit, I _think_ I should be hitting it here considering my low karma?
Screenshot 2023-04-26 at 12 22 00

- [ ] Button should ideally also be disabled also when there is no text in the comment box
- [ ] Saw you mentioned this on slack too–I think it's worth it having our own styling for disabled primary buttons (using 50% opacity)
- [ ] For dark mode the button should still have white text (I had a look around and see we don't do this almost anywhere on our blue buttons, but would be great to start doing it, see Githubs "comment" button as an example)
Screenshot 2023-04-26 at 12 15 35
Screenshot 2023-04-26 at 12 15 52


- [ ] For moderators, this dropdown is now kind of floating in the middle?
Screenshot 2023-04-26 at 12 07 01

**Comment by s-cheng** [comment]
> I never see the Cancel button 🤔 when should it show?

I believe this only appears in situations where the comment box is not open by default, so for example when you click "Reply" on a comment. This seems right to me?

> I also don't get the limit, I _think_ I should be hitting it here considering my low karma?

You should only see the warning and be rate limited after you've posted 4 comments within the last 30 min. Upon posting #4, the warning message should appear in the comment box, and the button should be disabled. What did you see that was wrong?

> Button should ideally also be disabled also when there is no text in the comment box

👍 

> Saw you mentioned this on slack too–I think it's worth it having our own styling for disabled primary buttons (using 50% opacity)

👍 

> For dark mode the button should still have white text (I had a look around and see we don't do this almost anywhere on our blue buttons, but would be great to start doing it, see Githubs "comment" button as an example)

Yeah I agree with this, though I might do this in a separate PR because I'd like to just fix them across the board.

> For moderators, this dropdown is now kind of floating in the middle?
>   Screenshot 2023-04-26 at 12 07 01

Yeah, though that's how it was before this as well - I'll see if there's a quick way to fix that, but prob not because it's part of the form input

**Comment by s-cheng** [comment]
@agnesstenlund I looked into disabling the button when there's no text, and unfortunately the way our forms are set up means that there's a delay (like a few seconds) in enabling it when you type. I think that's pretty bad, so I'm not going to implement this bit, but maybe we can look into it more in the future.

I also checked out the "Add comment" button on dark mode, and if I made the text white then it's a bit hard to read, because the background color is lighter on dark mode (it's #0c869b on light mode and #009da8 on dark mode). I only feel this a bit though, so if you want me to make this change I can keep it.
Screen Shot 2023-04-26 at 6 06 05 PM

File: packages/lesswrong/components/comments/CommentsNewForm.tsx
Comment (Will-Howard) [inline packages/lesswrong/components/comments/CommentsNewForm.tsx:1]: This looks so much better 😍

File: packages/lesswrong/components/events/modules/EventCards.tsx [hunk 2]
```diff
@@ -89,7 +89,7 @@ const styles = createStyles((theme: ThemeType): JssStyles => ({
   },
   eventCardGroup: {
     ...theme.typography.commentStyle,
-    ...theme.typography.italic,
+    fontStyle: "italic",
```
Comment (Will-Howard) [inline packages/lesswrong/components/events/modules/EventCards.tsx:92]: Reasonable decision in the circumstances

**Review by Will-Howard — APPROVED** [review]

**Comment by agnesstenlund** [comment]
> @agnesstenlund I looked into disabling the button when there's no text, and unfortunately the way our forms are set up means that there's a delay (like a few seconds) in enabling it when you type. I think that's pretty bad, so I'm not going to implement this bit, but maybe we can look into it more in the future.

seems reasonable 👍 

> I also checked out the "Add comment" button on dark mode, and if I made the text white then it's a bit hard to read, because the background color is lighter on dark mode (it's #0c869b on light mode and #009da8 on dark mode). I only feel this a bit though, so if you want me to make this change I can keep it.


Oh interesting, I hadn't realised. I'd probably use the same color for the button background in dark and light mode to get around this? I'll check what that looks like now

**Comment by agnesstenlund** [comment]
yup I actually think the slightly darker blue looks better in dark mode too:

Light mode blue:
Screenshot 2023-04-27 at 20 55 13

Dark mode blue:
Screenshot 2023-04-27 at 20 55 31

**Comment by agnesstenlund** [comment]
> You should only see the warning and be rate limited after you've posted 4 comments within the last 30 min. Upon posting https://github.com/ForumMagnum/ForumMagnum/issues/4, the warning message should appear in the comment box, and the button should be disabled. What did you see that was wrong?

Oh I've tried a few times posting more than 4, and nothing has showed up 😮  could it be because I'm a moderator? Even though I've low karma?

**Comment by agnesstenlund** [comment]
also found this other piece of information in the comment box when logged out! If not huge to do we should use the same info banner in this case 🙌  but with this text instead `You have be logged in to comment`
Screenshot 2023-04-27 at 21 04 33

**Comment by agnesstenlund** [comment]
Hmm so I don't know if this is just happening for me–but this showed for me twice before actually hitting 4 comments:
Screenshot 2023-04-27 at 21 07 57

it goes away when I reload and then I can comment on another post

**Comment by s-cheng** [comment]
@agnesstenlund 
> Oh I've tried a few times posting more than 4, and nothing has showed up 😮 could it be because I'm a moderator? Even though I've low karma?

Oh sorry, I didn't notice that LW had added this change since I opened this PR. Yeah, admins and mods are exempt from the low karma rate limit. Currently the post author is also exempt, though I'm not sure if we'll keep that bit.

> also found this other piece of information in the comment box when logged out!

I updated that to use the yellow warning banner 👍 

> Hmm so I don't know if this is just happening for me–but this showed for me twice before actually hitting 4 comments

Yeah apparently we have a universal rate limit on commenting, so that users don't accidentally double submit. But it was set to 15 sec, which I think is too long - I've reduced it to 5 sec, so I think users shouldn't really run into it anymore.

**Comment by s-cheng** [comment]
@agnesstenlund I think I've addressed all your comments now, let me know if I missed anything!

**Comment by s-cheng** [comment]
> I got this now when trying to post two comments within a short space of time, could that be because of the 5 sec limit of what's happening here 🤔

Oh gosh, I dunno how you ended up there! I can't seem to repro this but maybe we should increase the universal rate limit a bit just in case. I'll make it 8 sec? 🤷‍♀️

**Pull request merged**